### ISSUE TITLE
Exception during Migration

### ISSUE THREAD

**Issue by Sutirth** [realm/realm-java#2534]
#### Goal

Remove certain columns, Adding columns, renaming them.

#### Expected Results

Should be able migrate 

#### Actual Results
```
03-31 13:52:03.901 9535-9535/com.workoutguru.workoutcash E/AndroidRuntime: Process: com.workoutguru.workoutcash, PID: 9535
03-31 13:52:03.901 9535-9535/com.workoutguru.workoutcash E/AndroidRuntime: io.realm.exceptions.RealmMigrationNeededException: Primary key not defined for field 'userId' in existing Realm file. Add @PrimaryKey.
03-31 13:52:03.901 9535-9535/com.workoutguru.workoutcash E/AndroidRuntime:     at io.realm.LeaderBoardDbRealmProxy.validateTable(LeaderBoardDbRealmProxy.java:337)
03-31 13:52:03.901 9535-9535/com.workoutguru.workoutcash E/AndroidRuntime:     at io.realm.DefaultRealmModuleMediator.validateTable(DefaultRealmModuleMediator.java:121)
```

#### Code Sample
```java
public class LeaderBoardDb extends RealmObject {
//    private String percentile;
//
//    private String rank;

    private String details;

    private String email;

    @PrimaryKey
    private String userId;

    private String name;

//    private String rewards;
//
//    private String sweatPointEarned;

    private String photoUrl;

    private String note;

    private String displayName;

    private float userPercentile;

    private int userRank;

    private int userRewards;

    private int userSweatPointsEarned;
}
```
```java
if(oldVersion == 6){


                schema.get("LeaderBoardDb")
                        .addField("userPercentile", float.class)
                .addField("userRank",int.class)
                .addField("userRewards",int.class)
                .addField("userSweatPointsEarned", int.class)
                .removeField("percentile")
                .removeField("rank")
                .removeField("rewards")
                .removeField("sweatPointEarned")
                        .renameField("userPercentile", "percentile")
                        .renameField("userRank", "rank")
                        .renameField("userRewards","rewards")
                        .renameField("userSweatPointsEarned", "sweatPointEarned");

                oldVersion++;
            }
```

My primary key in the database field already exist in earlier version commented columns are the name

**Comment by cmelchior** [comment]
Hi @Sutirth 
Are you able to create a small test project that showcase this error. If you want to keep it private you can send it to help@realm.io ?

**Comment by Sutirth** [comment]
Hi @cmelchior 

I am extremely sorry for creating issue at 2 places. But I am in situation where my project crashes and I am unable to do anything. 

I am actually having time crunches at the moment. Could tell me a way to fix this ?? or any other possible ways

**Comment by cmelchior** [comment]
@Sutirth The exception message you are getting does make sense considering the changes you made, so I suspect something else is going on that we cannot see from the source code you provided.

We need some way to reproduce this if we are going to help you, so if you can create an example project or similar that showcase the error that will be the fastest.

**Comment by Sutirth** [comment]
public class Migration implements RealmMigration {

        @Override
        public void migrate(final DynamicRealm realm, long oldVersion, long newVersion){

            RealmSchema schema = realm.getSchema();

            //  ***********************************************************
//  **** This will migrate from version 0 to version 1 ****
//  ***********************************************************
            if (oldVersion == 0) {
                RealmObjectSchema sweatPointSchema = schema.get("UserInfoDB");
                sweatPointSchema.addField("sweatPoints",Integer.class);
                oldVersion++;
            }

//  ***********************************************************
//  **** This will migrate from version 1 to version 2 ****
//  ***********************************************************
            if(oldVersion == 1) {
                RealmObjectSchema teamInfoSchema = schema.get("TeamInfoDB");
                teamInfoSchema.addField("teamDescription",String.class);

                oldVersion ++;
            }

//  ***********************************************************
//  **** This will migrate from version 2 to version 3 ****
//  ***********************************************************
            if(oldVersion == 2) {
                RealmObjectSchema leaderboardInfoSchema = schema.get("LeaderBoardDb");
                leaderboardInfoSchema.addField("displayName",String.class);



                oldVersion ++;
            }
//  ***********************************************************
//  **** This will migrate from version 3 to version 4 ****
//  ***********************************************************

            if(oldVersion == 3){

                RealmObjectSchema sweatPointFitnessSchema = schema.get("SweatPointFitnessDB");
                sweatPointFitnessSchema.addField("badgeSweatPoints", int.class);
                sweatPointFitnessSchema.setNullable("badgeSweatPoints", true);


                RealmObjectSchema userInfoSchema = schema.get("UserInfoDB");
                userInfoSchema.addField("profileImageUpdatedAt",String.class);

                oldVersion++;
            }


            //  ***********************************************************
//  **** This will migrate from version 4 to version 5 ****
//  ***********************************************************


            if(oldVersion == 4){

                RealmObjectSchema sweatPointFitnessSchema = schema.get("SweatPointFitnessDB");
                sweatPointFitnessSchema.setNullable("badgeSweatPoints", false);




                oldVersion++;
            }

            //  ***********************************************************
//  **** This will migrate from version 5 to version 6 ****
//  ***********************************************************

            if(oldVersion == 5){

                RealmObjectSchema sweatPointFitnessSchema = schema.get("SweatPointFitnessDB");
                sweatPointFitnessSchema.setNullable("badgeSweatPoints", true);




                oldVersion++;
            }

            //  ***********************************************************
            //  **** This will migrate from version 6 to version 7 ****
            //  ***********************************************************

            if(oldVersion == 6){


                schema.get("LeaderBoardDb")
                .addField("userPercentile", float.class)
                .addField("userRank", int.class)
                .addField("userRewards",int.class)
                .addField("userSweatPointsEarned", int.class)
                .removeField("percentile")
                .removeField("rank")
                .removeField("rewards")
                .removeField("sweatPointEarned")
                        .renameField("userPercentile", "percentile")
                        .renameField("userRank", "rank")
                        .renameField("userRewards","rewards")
                        .renameField("userSweatPointsEarned", "sweatPointEarned");

                oldVersion++;
            }
}
This is my whole source code for migration. 
@cmelchior  Anything else I need to share I don't mind sharing code

**Comment by stk1m1** [comment]
@Sutirth Thanks for sharing your code. We'll try to reproduce your troubled case, and contact you soon.

**Comment by cmelchior** [comment]
Hi @Sutirth 
There is nothing in your migration code that should indicate the error you are getting. Are you able to create a demo project that fails the migration?

**Comment by Sutirth** [comment]
Hi @cmelchior  In demo project the migration does not fails. Do you think may be I am using compile 'io.realm:realm-android:0.87.5' this might contain some bug

**Comment by cmelchior** [comment]
I assume your demo project uses 0.87.5 as well? Also remember to include an old realm file that needs to be updated, otherwise the Realm migration won't be triggered.

**Comment by Sutirth** [comment]
@cmelchior  Yes you right I am following all procedures as per protocols

**Comment by cmelchior** [comment]
Hi @Sutirth 
Have you been able to create a demo project that shows the error?

**Comment by kneth** [comment]
@Sutirth Any progress on reproducing the error in a small project?

**Comment by biagionicp** [comment]
@cmelchior @kneth I experienced a similar error. I think I may know what caused it. When updating my realm schema, i reconfigured the `RealmMigration`  to add the new class/fields , however I forgot to update the schema version. When I updated the app, I experienced no issues, but then realized I forgot to update the schema version, so when I updated it, I got the class already exists exception which was confusing because I wouldn't think the migration would have occured without updating the schema version, but apparently it did.
Is it possible that realm can execute a migration with the given `RealmMigration` object if it experiences a realm migration exception; even if the schema version was never updated?
